ipfshttpclient/filescanner.py

```diff
@@ -728,12 +734,8 @@ def _walk(
 						)
 		finally:
 			# Make sure the file descriptors bound by `os.fwalk` are freed on error
-			walk_iter.close()  # type: ignore[union-attr]  # typeshed bug
-			
-			# Close root file descriptor of `os.fwalk` as well
-			if self._close_fd is not None:
-				os.close(self._close_fd)
-				self._close_fd = None
```

Comment:
Neither `os.walk` nor `os.fwalk` declare a return type that has a `close` method.

Reply:
Here's the implementation of `os.fwalk`:

    def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=None):
        """Directory tree generator.

        This behaves exactly like walk(), except that it yields a 4-tuple

            dirpath, dirnames, filenames, dirfd

        `dirpath`, `dirnames` and `filenames` are identical to walk() output,
        and `dirfd` is a file descriptor referring to the directory `dirpath`.

        The advantage of fwalk() over walk() is that it's safe against symlink
        races (when follow_symlinks is False).

        If dir_fd is not None, it should be a file descriptor open to a directory,
          and top should be relative; top will then be relative to that directory.
          (dir_fd is always supported for fwalk.)

        Caution:
        Since fwalk() yields file descriptors, those are only valid until the
        next iteration step, so you should dup() them if you want to keep them
        for a longer period.

        Example:

        import os
        for root, dirs, files, rootfd in os.fwalk('python/Lib/email'):
            print(root, "consumes", end="")
            print(sum(os.stat(name, dir_fd=rootfd).st_size for name in files),
                  end="")
            print("bytes in", len(files), "non-directory files")
            if 'CVS' in dirs:
                dirs.remove('CVS')  # don't visit CVS directories
        """
        if not isinstance(top, int) or not hasattr(top, '__index__'):
            top = fspath(top)
        # Note: To guard against symlink races, we use the standard
        # lstat()/open()/fstat() trick.
        if not follow_symlinks:
            orig_st = stat(top, follow_symlinks=False, dir_fd=dir_fd)
        topfd = open(top, O_RDONLY, dir_fd=dir_fd)
        try:
            if (follow_symlinks or (st.S_ISDIR(orig_st.st_mode) and
                                    path.samestat(orig_st, stat(topfd)))):
                yield from _fwalk(topfd, top, isinstance(top, bytes),
                                  topdown, onerror, follow_symlinks)
        finally:
            close(topfd)

Notice it `close`s itself when the generator is completed. This is why the new implementation has a `for` loop with a `pass` body in `finally`

Reply:
Every generator has `.close()`, `.send(…)` and `.throw(…)` methods implicitly. Calling `.close()` on the generator will raise a `GeneratorExit` exception at the `yield` that the generator is currently blocked on. Any `yield from`s will propagate their received exceptions and values “down-wards” into the generators *they* are blocked on. See [PEP-342](https://www.python.org/dev/peps/pep-0342/) for details on the “new” since Python 2.5 generators, that are more than simple iterables and [PEP-380](https://www.python.org/dev/peps/pep-0380/#formal-semantics) for the instructive definition of `yield from`.

Reply:
Thanks for the additional info. I'll take a closer look at this again.

Reply:
I've made an update that addresses this. It deduplicates the file descriptor closing (I found another identical block of code).

Out of curiosity, why is walk() a generator class instead of a generator function?

Reply:
> I've made an update that addresses this. It deduplicates the file descriptor closing (I found another identical block of code).

Thank you! Never noticed that!

> Out of curiosity, why is walk() a generator class instead of a generator function?

You got me there for a second! :slightly_smiling_face: 

As far as I can tell, it's so that it can act as both a generator and a context-manager. So the following works:

```py
with filescanner.walk("./bla") as scanner:
    for item in scanner:
        process_item(item)  # Something that could raise
```

Instead of more awkward:

```py
scanner = filescanner.walk("./bla")
try:
    for item in scanner:
        process_item(item)  # Something that could raise
finally:
    scanner.close()
```

Reply:
The former is much nicer than the latter.